### DISCUSSION TITLE
Websocket Server with gnet

### DISCUSSION THREAD

**Discussion by spy16** [panjf2000/gnet#520] [Q&A]
I am trying to use `gnet` for building a high throughput websocket server.

I tried to directly use the `gobwas/ws` and `gobwas/ws/wsutil` packages to read from and write to connections using `wsutil.WriteServerMessage()` and `wsutil.ReadClientText`. I don't think this is the correct way? In the [gnet examples](https://github.com/gnet-io/gnet-examples/blob/v2/websocket/server), I see there is some codec implementation that is used in conjunction with `gobwas/ws` packages, is that a necessary or is there any easier way?

**Comment by panjf2000** [comment]
There are no ***right*** patterns and standards of developing WebSocket frameworks using `gnet`, just keep in mind that never do anything blocking (such as network I/O) in `OnTraffic`.

**Reply by spy16x** [reply]
Hey @panjf2000 !! .. to clarify I'm not looking for the best patterns or anything here. Just trying to understand if my understanding is fundamentally incorrect in terms of how gnet and ws work together.

I built a version very similar to the one here: 
 https://github.com/nubunto/gnet-websocket-broadcast/blob/main/websocket.go

With one difference: after doing `wsutil.ReadClientData` I launched a goroutine using the ants pool so as to not block the IO loop. 

I tried load testing this and I ended up receiving a few bad file descriptor errors and a panic with concurrent map write during `wsutil.WriteServerMessage`. On client side (`websocat` cli) also reported malformed data receive. 

In the gnet-examples however, along with wsutil, there is also a codec being used. I'm not entirely sure what is need for that.

**Reply by panjf2000** [reply]
Do you have any stack info for your errors? It's hard for me to diagnose some issues without info.

**Reply by spy16x** [reply]
I realised that the malformed data issue is related to this: https://github.com/gnet-io/gnet-examples/issues/16

not sure about the bad file descriptor and panic. i will share details if i run my load testing again.
